services/factorio-mod-portal/factorio-mod-portal.service.js

```diff
@@ -0,0 +1,198 @@
+import { renderVersionBadge } from '../version.js'
+
+const schema = Joi.object({
+  downloads_count: Joi.number().required(),
+  releases: Joi.array()
+    .items(
+      Joi.object({
+        version: Joi.string().required(),
+        released_at: Joi.string().required(),
+        info_json: Joi.object({
+          factorio_version: Joi.string().required(),
+        }).required(),
+      })
+    )
+    .min(1)
+    .required(),
+}).required()
+
+const queryParamSchema = Joi.object({
+  range: Joi.equal(''),
+}).required()
+
+// Factorio Mod portal API
+// @see https://wiki.factorio.com/Mod_portal_API
+class BaseFactorioModPortalService extends BaseJsonService {
+  async fetch({ modName }) {
+    const { releases, downloads_count } = await this._requestJson({
+      schema,
+      url: `https://mods.factorio.com/api/mods/${modName}`,
+      errorMessages: {
+        404: 'mod not found',
+      },
+    })
+
+    return {
+      downloads_count,
+      first_release: releases[0],
+      latest_release: releases[releases.length - 1],
+    }
+  }
+}
+
+// Badge for mod's latest updated version
+class FactorioModPortalLatestVersion extends BaseFactorioModPortalService {
+  static category = 'version'
+
+  static route = {
+    base: 'factorio-mod-portal/latest-version',
+    pattern: ':modName',
+  }
+
+  static examples = [
+    {
+      title: 'Factorio Mod Portal mod version',
+      namedParams: { modName: 'rso-mod' },
+      staticPreview: this.render({ version: '6.2.20' }),
+    },
+  ]
+
+  static defaultBadgeData = { label: 'latest version' }
+
+  static render({ version }) {
+    return renderVersionBadge({ version })
+  }
+
+  async handle({ modName }) {
+    const { latest_release } = await this.fetch({ modName })
+    return this.constructor.render({ version: latest_release.version })
+  }
+}
+
+// Badge for mod's latest compatible Factorio version
+// Query 'range' to display range of compatible versions
+class FactorioModPortalFactorioVersions extends BaseFactorioModPortalService {
+  static category = 'version'
+
+  static route = {
+    base: 'factorio-mod-portal/factorio-version',
+    pattern: ':modName',
+    queryParamSchema,
+  }
+
+  static examples = [
+    {
+      title: 'Factorio Mod Portal factorio versions',
+      namedParams: { modName: 'rso-mod' },
+      staticPreview: this.render({ version: '0.14-1.1' }),
+      queryParams: {
+        range: 'range',
+      },
+    },
+  ]
+
+  static defaultBadgeData = { label: 'factorio version' }
+
+  static render({ version }) {
+    return { message: version, color: 'blue' }
+  }
+
+  transform({ earliest, latest }) {
+    let versions = ''
+    if (earliest === latest) {
+      versions = earliest
+    } else {
+      versions = `${earliest}-${latest}`
+    }
+    return versions
+  }
+
+  async handle({ modName }, queryParams) {
+    const { first_release, latest_release } = await this.fetch({ modName })
+    const range = queryParams.range !== undefined
```

Comment:
Generally for platform & version support badges, we just show compatibility with the base platform for the latest release of the target package/plugin/whatever (in this case, the factorio version(s) that the latest release of the mod is compatible with). I think I'd be in favour of dropping the `range` param and sticking with that here.